File: backend/app/api/v1/vehicles.py
"""
차량 마스터 조회 API 엔드포인트
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional

from app.core.database import get_db
from app.schemas.vehicle import (
    StandardResponse,
    ManufacturerResponse,
    ModelGroupResponse,
    VehicleModelDetailResponse,
    VehicleClassResponse
)
from app.services.vehicle_service import VehicleService

router = APIRouter(prefix="/vehicles", tags=["차량 마스터"])


@router.get("/manufacturers", response_model=StandardResponse)
async def get_manufacturers(
    origin: Optional[str] = Query(None, description="국산/수입 구분 (domestic, imported)"),
    db: AsyncSession = Depends(get_db)
):
    """
    제조사 목록 조회 API
    
    - 국산/수입 구분 필터링 지원
    - Redis 캐싱 적용 (TTL: 1시간)
    """
    if origin and origin not in ["domestic", "imported"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="origin 파라미터는 'domestic' 또는 'imported'여야 합니다"
        )
    
    manufacturers = await VehicleService.get_manufacturers(db, origin)
    
    return StandardResponse(
        success=True,
        data=manufacturers,
        error=None
    )


@router.get("/model-groups", response_model=StandardResponse)
async def get_model_groups(
    manufacturer: str = Query(..., description="제조사명"),
    origin: Optional[str] = Query(None, description="국산/수입 구분 (domestic, imported)"),
    db: AsyncSession = Depends(get_db)
):
    """
    모델 그룹 목록 조회 API
    
    - 제조사별 모델 그룹 조회
    - Redis 캐싱 적용 (TTL: 1시간)
    """
    if origin and origin not in ["domestic", "imported"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="origin 파라미터는 'domestic' 또는 'imported'여야 합니다"
        )
    
    model_groups = await VehicleService.get_model_groups(db, manufacturer, origin)
    
    return StandardResponse(
        success=True,
        data=model_groups,
        error=None
    )


@router.get("/models", response_model=StandardResponse)
async def get_models(
    manufacturer: Optional[str] = Query(None, description="제조사명"),
    model_group: Optional[str] = Query(None, description="모델 그룹명"),
    origin: Optional[str] = Query(None, description="국산/수입 구분 (domestic, imported)"),
    db: AsyncSession = Depends(get_db)
):
    """
    차량 모델 목록 조회 API
    
    - 제조사/모델 그룹/국산수입 구분 필터링 지원
    - Redis 캐싱 적용 (TTL: 1시간)
    """
    if origin and origin not in ["domestic", "imported"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="origin 파라미터는 'domestic' 또는 'imported'여야 합니다"
        )
    
    models = await VehicleService.get_models(db, manufacturer, model_group, origin)
    
    return StandardResponse(
        success=True,
        data=models,
        error=None
    )


@router.get("/models/{model_id}/details", response_model=StandardResponse)
async def get_model_details(
    model_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    차량 모델 상세 정보 조회 API
    
    - 모델 ID로 상세 정보 조회
    - 연식, 배기량 등 상세 정보 포함
    """
    model_details = await VehicleService.get_model_details(db, model_id)
    
    if not model_details:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="차량 모델을 찾을 수 없습니다"
        )
    
    return StandardResponse(
        success=True,
        data=model_details,
        error=None
    )


@router.get("/classes", response_model=StandardResponse)
async def get_vehicle_classes(
    origin: Optional[str] = Query(None, description="국산/수입 구분 (domestic, imported)"),
    db: AsyncSession = Depends(get_db)
):
    """
    차량 등급 목록 조회 API
    
    - 경차/소형/중형/대형/SUV/스포츠카/슈퍼카 등급 정보
    - 국산/수입 구분 필터링 지원
    """
    if origin and origin not in ["domestic", "imported"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="origin 파라미터는 'domestic' 또는 'imported'여야 합니다"
        )
    
    classes = await VehicleService.get_vehicle_classes(db, origin)
    
    return StandardResponse(
        success=True,
        data=classes,
        error=None
    )


@router.get("/lookup", response_model=StandardResponse)
async def lookup_vehicle_by_plate(
    plate_number: str = Query(..., description="차량번호"),
    db: AsyncSession = Depends(get_db)
):
    """
    차량번호 기반 조회 API (국토교통부 API 연동 기초 작업)
    
    - 차량번호로 차량 정보 조회
    - 국토교통부 API 연동 예정 (현재는 기본 구조만 구현)
    """
    # TODO: 국토교통부 API 연동 구현
    # 현재는 기본 응답만 반환
    
    return StandardResponse(
        success=True,
        data={
            "plate_number": plate_number,
            "message": "국토교통부 API 연동 예정"
        },
        error=None
    )

File: frontend/lib/api/vehicles.ts
/**
 * 차량 마스터 조회 API 클라이언트
 * 고객 신청 플로우에서 사용
 */
import apiClient from './client';
import type { StandardResponse } from '@/types/api';

// 제조사 정보
export interface Manufacturer {
  name: string;
  origin: 'domestic' | 'imported';
}

// 모델 그룹 정보
export interface ModelGroup {
  name: string;
  manufacturer: string;
}

// 차량 모델 정보
export interface VehicleModel {
  id: string;
  manufacturer: string;
  model_group: string;
  model_detail?: string | null;
  vehicle_class: string;
  origin: 'domestic' | 'imported';
}

// 차량 모델 상세 정보
export interface VehicleModelDetail {
  id: string;
  manufacturer: string;
  model_group: string;
  model_detail?: string | null;
  vehicle_class: string;
  origin: 'domestic' | 'imported';
  start_year: number;
  end_year?: number | null;
}

// 차량번호 조회 응답
export interface VehicleLookupResponse {
  plate_number: string;
  message?: string;
  // TODO: 국토교통부 API 연동 후 실제 차량 정보 추가
}

/**
 * 제조사 목록 조회
 */
export const getManufacturers = async (origin?: 'domestic' | 'imported'): Promise<Manufacturer[]> => {
  try {
    const params = origin ? { origin } : {};
    const response = await apiClient.get<StandardResponse<Manufacturer[]>>('/vehicles/manufacturers', { params });
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '제조사 목록을 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('제조사 목록 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('제조사 목록을 불러올 수 없습니다');
  }
};

/**
 * 모델 그룹 목록 조회
 */
export const getModelGroups = async (
  manufacturer: string,
  origin?: 'domestic' | 'imported'
): Promise<ModelGroup[]> => {
  try {
    const params: any = { manufacturer };
    if (origin) {
      params.origin = origin;
    }
    const response = await apiClient.get<StandardResponse<ModelGroup[]>>('/vehicles/model-groups', { params });
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '모델 그룹 목록을 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('모델 그룹 목록 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('모델 그룹 목록을 불러올 수 없습니다');
  }
};

/**
 * 차량 모델 목록 조회
 */
export const getVehicleModels = async (
  manufacturer?: string,
  model_group?: string,
  origin?: 'domestic' | 'imported'
): Promise<VehicleModel[]> => {
  try {
    const params: any = {};
    if (manufacturer) params.manufacturer = manufacturer;
    if (model_group) params.model_group = model_group;
    if (origin) params.origin = origin;
    
    const response = await apiClient.get<StandardResponse<VehicleModel[]>>('/vehicles/models', { params });
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '차량 모델 목록을 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('차량 모델 목록 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('차량 모델 목록을 불러올 수 없습니다');
  }
};

/**
 * 차량 모델 상세 정보 조회
 */
export const getVehicleModelDetail = async (modelId: string): Promise<VehicleModelDetail> => {
  try {
    const response = await apiClient.get<StandardResponse<VehicleModelDetail>>(`/vehicles/models/${modelId}/details`);
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '차량 모델 상세 정보를 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('차량 모델 상세 정보 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('차량 모델 상세 정보를 불러올 수 없습니다');
  }
};

/**
 * 차량번호 조회
 */
export const lookupVehicleByPlate = async (plateNumber: string): Promise<VehicleLookupResponse> => {
  try {
    const response = await apiClient.get<StandardResponse<VehicleLookupResponse>>('/client/vehicle/lookup', {
      params: { plate_number: plateNumber },
    });
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '차량번호 조회에 실패했습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('차량번호 조회 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('차량번호 조회에 실패했습니다');
  }
};

